src/user_lists/serializers.py

```diff
@@ -1,20 +1,157 @@
+
+    def _get_content(self, document):
+        if hasattr(document, "_prefetched_objects_cache") and "posts" in document._prefetched_objects_cache:
+            posts = document._prefetched_objects_cache["posts"]
+            if posts:
+                return posts[0]
+        
+        try:
+            return document.paper
+        except (ObjectDoesNotExist, AttributeError):
+            return None
+
+    def _create_feed_entry(self, content, document):
+        if not content or not hasattr(content, 'id'):
+            return None
+            
+        content_type_cache = self.context.get('content_type_cache', {})
+        content_class = content.__class__
+        content_type = content_type_cache.get(content_class)
+        
+        if not content_type:
+            return None
+        
+        metrics_key = (content_class, content.id)
+        if metrics_key not in self._metrics_cache:
+            self._metrics_cache[metrics_key] = serialize_feed_metrics(content, content_type)
+
+        entry = FeedEntry(
+            id=content.id,
+            content_type=content_type,
+            object_id=content.id,
+            action="PUBLISH",
+            action_date=getattr(content, "paper_publish_date", None) or getattr(content, "created_date", None),
+            created_date=getattr(content, "created_date", None),
+            user=getattr(content, "created_by", None) or getattr(content, "uploaded_by", None),
+            unified_document=document,
+            hot_score_v2=getattr(content, "hot_score_v2", 0),
+        )
+        entry.item = content
+        entry.metrics = self._metrics_cache[metrics_key]
+        return entry
+
+
+class ListItemWriteSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = ListItem
+        fields = LIST_ITEM_WRITABLE_FIELDS
+
+    def validate_parent_list(self, value):
+        user = self.context["request"].user
+        if value.created_by != user or value.is_removed:
+            raise serializers.ValidationError("List not found or you don't have permission.")
+        return value
+
+    def validate(self, attrs):
+        if self.instance:
+            parent_list = attrs.get("parent_list", self.instance.parent_list)
+            unified_document = attrs.get("unified_document", self.instance.unified_document)
+            
+            if parent_list != self.instance.parent_list or unified_document != self.instance.unified_document:
+                self._validate_no_duplicate_item(parent_list, unified_document)
+        return attrs
+
+    def _validate_no_duplicate_item(self, parent_list, unified_document):
+        if ListItem.objects.filter(
+            parent_list=parent_list,
+            unified_document=unified_document,
+            is_removed=False
+        ).exclude(id=self.instance.id if self.instance else None).exists():
+            raise serializers.ValidationError("Item already exists in this list.")
```

Comment:
What will happen when this is raised? what would the frontend receive? object? or string?

Reply:
I see sometimes we add "error": and other times we just return string